libraries/Bluefruit52Lib/src/services/BLEDis.cpp

```diff
@@ -51,51 +51,164 @@ BLEDis::BLEDis(void)
+void BLEDis::setFirmwareRev(const char* firmware_rev, uint8_t length)
+{
+  _firmware_rev = firmware_rev;
+  _firmware_rev_length = length;
+}
+
+void BLEDis::setSerialNum(const char* serial_num)
+{
+  _serial = serial_num;
+  _serial_length = strlen(serial_num);
+}
+void BLEDis::setSerialNum(const char* serial_num, uint8_t length)
+{
+  _serial = serial_num;
+  _serial_length = length;
+}
+
+void BLEDis::setSystemID(const char* system_id)
+{
+  _system_id = system_id;
+  _system_id_length = strlen(system_id);
+}
+void BLEDis::setSystemID(const char* system_id, uint8_t length)
+{
+  _system_id = system_id;
+  _system_id_length = length;
+}
+
+void BLEDis::setRegCertList(const char* reg_cert_list)
+{
+  _reg_cert_list = reg_cert_list;
+  _reg_cert_list_length = strlen(reg_cert_list);
+}
+void BLEDis::setRegCertList(const char* reg_cert_list, uint8_t length)
+{
+  _reg_cert_list = reg_cert_list;
+  _reg_cert_list_length = length;
+}
+
+void BLEDis::setPNPID(const char* pnp_id)
+{
+  _pnp_id = pnp_id;
+  _pnp_id_length = strlen(pnp_id);
+}
+void BLEDis::setPNPID(const char* pnp_id, uint8_t length)
+{
+  _pnp_id = pnp_id;
+  _pnp_id_length = length;
+}
+void BLEDis::setModel(const char* model,uint8_t length)
+{
+  _model = model;
+  _model_length = length;
+}
+
+void BLEDis::setHardwareRev(const char* hw_rev,uint8_t length)
+{
+  _hardware_rev = hw_rev;
+  _hardware_rev_length = length;
+}
+
+void BLEDis::setSoftwareRev(const char* sw_rev, uint8_t length)
+{
+  _software_rev = sw_rev;
+  _software_rev_length = length;
+}
+
+void BLEDis::setManufacturer(const char* manufacturer, uint8_t length)
+{
+  _manufacturer = manufacturer;
+  _manufacturer_length = length;
 }
 
 void BLEDis::setModel(const char* model)
 {
   _model = model;
+  _model_length = strlen(model);
 }
 
 void BLEDis::setHardwareRev(const char* hw_rev)
 {
   _hardware_rev = hw_rev;
+  _hardware_rev_length = strlen(hw_rev);
 }
 
 void BLEDis::setSoftwareRev(const char* sw_rev)
 {
   _software_rev = sw_rev;
+  _software_rev_length = strlen(sw_rev);
 }
 
 void BLEDis::setManufacturer(const char* manufacturer)
 {
   _manufacturer = manufacturer;
+  _manufacturer_length = strlen(manufacturer);
 }
 
 err_t BLEDis::begin(void)
 {
   // Invoke base class begin()
   VERIFY_STATUS( BLEService::begin() );
 
-  _serial       = getMcuUniqueID();
-  _firmware_rev = getBootloaderVersion();
+  if(!_serial) {
+    _serial       = getMcuUniqueID();
+  }
+  if (!_firmware_rev) {
+    _firmware_rev = getBootloaderVersion();
+  }
 
-  for(uint8_t i=0; i<arrcount(_strarr); i++)
+  for(uint8_t i=0; i<arrcount(_strarr)-1; i++)    // without PNP_ID
   {
     if ( _strarr[i] != NULL )
     {
-      BLECharacteristic chars(UUID16_CHR_MODEL_NUMBER_STRING+i);
+      BLECharacteristic chars(UUID16_CHR_SYSTEM_ID+i);
       chars.setTempMemory();
-
       chars.setProperties(CHR_PROPS_READ);
-      chars.setFixedLen(strlen(_strarr[i]));
-
+      if (_strarr_length[i]) {
+        chars.setFixedLen(_strarr_length[i]);
```

Comment:
3. should only use the _strarr_legnth[i], there is no point to use the strlen() function anymore.

Reply:
Yes, you are rigth.